Bug 1966499: Switch Cypress OLM tests to use supported Red Hat operators

dtaylor113 opened the pull request:
- OLM single-install test changed from community Couchbase operator to Red Hat CodeReady Workspaces
- OLM global-install test changed from community Portworx operator to Red Hat Service Binding Operator

I replaced the entire contents of `operator-install-global.spec.ts` and `operator-install-single-namespace.spec.ts` so I suggest looking at these files in 'raw format' as the Diff view is hard to follow due to so many changes.

TheRealJon commented on frontend/packages/operator-lifecycle-manager/integration-tests-cypress/views/operator.ts:
```diff
@@ -0,0 +1,144 @@
+import { nav } from '@console/cypress-integration-tests/views/nav';
+
+export const operator = {
+  install: (
```
The install action should be decoupled from the assertion. What if we want to write a test case for an operator that should fail to install?

dtaylor113 replied:
Please see the use of `cy.intercept()` in [#9142](https://github.com/openshift/console/pull/9142/files#diff-888fcd3f03658ad6355c60586b25467f9f1ca3f95c21fc6869bb2b4f3531a2a3R104) to mock failed uninstallation.

I think it would be rather difficult to decouple assertions from commands in this case since we kinda want to test assertions as we go through the many steps of installing an operator or creating an operand instance.   At the moment we have two tests (OLM install to single and global namespaces), it we took out the assertions, they would have to be duplicated in each test suite which uses the operator view.

TheRealJon replied:
Right, you can mock a failure, but you would still need to either create a new helper or just write the test case manually for that. You wouldn't be able to use your `operator.install` helper at all in order to test a failed install. IMO, it would be better if our helpers were less coupled to specific test cases so that they could be reused. If you wrote an install/uninstall helper that is agnostic to success or failure, then it could be reused in the story that you linked.

TheRealJon commented on frontend/packages/operator-lifecycle-manager/integration-tests-cypress/views/operator.ts:
```diff
@@ -0,0 +1,144 @@
+    operatorName: string,
+    operatorHubCardID: string,
+    installToNamespace: string,
+    installToSingleNamespace: boolean = false,
```
This argument isn't really necessary. We can just take 3 args, and if the `installToNamespace` arg is undefined, we assume a global install and use the 'openshift-operators' namespace.

dtaylor113 replied:
I did have it that way, but openned it up to be more consistant between single and global installs, and provide maximum flexibility -wasn't sure if it always `openshit-operators` for every Operator (I know it is for RH Data Grid we're using, and probably for others, but don't know if that will always 100% of the time be the case).

TheRealJon replied:
In current OLM, it will always be 'openshift-operators'. That's a system namespace that's specifically there for globally installed operators.

TheRealJon commented on frontend/packages/operator-lifecycle-manager/integration-tests-cypress/views/operator.ts:
```diff
@@ -0,0 +1,144 @@
+    checkDeleteAllOperands: () =>
+      cy.byTestID('Delete all operand instances for this operator__checkbox').click(),
+  },
+  createInstance: (
```
Separate CRUD action from assertion so that we aren't tied to a specific post-action state.
We should be able to do something like:

```
operator.createOperand(...)
operator.operandExists(...)
```

dtaylor113 replied:
We do assertions along-the-way as we do these OLM ops. 
For example for `operator.install()`
1. assert global or single  'installedNamespace` is in the ns dropdown on the install form
2. click Install button
3. assert 'Operator begun installation' messages
4. click 'view installed operator' button
5. assert operator appears in Installed Operators list
6. assert operator's status reaches 'Succeeded'.

I can't really take out/decouple the assertions from this flow.

TheRealJon replied:
I just mean for the final resulting state. I think it's okay to assert certain states along the way, but I think it would be better if the `install` helper was agnostic to whether or not the installation should succeed or not.

TheRealJon replied:
Same for the create action, it should be agnostic as to whether or not the creation was successful. That should be up to the test case to assert.

dtaylor113 replied:
Same for `operator.createInstance/Operand()`:
1. assert initial "No operands found" messages
2. nav to creation form
3. assert correct form title
4. type operand link name
5. click create button
6. assert link name exists in table
7. click on assert link
8. assert operand details page title

dtaylor113 replied:
> I just mean for the final resulting state. I think it's okay to assert certain states along the way...

Ok, I'll more out the last assertions from these commands, and add appropriate `shouldExist` helpers :-)

TheRealJon commented on frontend/packages/operator-lifecycle-manager/integration-tests-cypress/views/operator.ts:
```diff
@@ -0,0 +1,144 @@
+    cy.byTestOperandLink(linkName).click();
+    cy.byTestSectionHeading(`${name} overview`).should('exist');
+  },
+  deleteInstance: (
```
Decouple action/assertion

TheRealJon commented on frontend/packages/operator-lifecycle-manager/integration-tests-cypress/views/operator.ts:
```diff
@@ -0,0 +1,144 @@
+      'Operands are declarative components used to define the behavior of the application.',
+    );
+  },
+  uninstall: (operatorName: string, installedNamespace: string) => {
```
Decouple action/assertion

TheRealJon reviewed (commented):
A couple of nits and a request to decouple our CRUD actions from test assertions. It's okay to have assertions for intermediate state (like form inputs), but I think after the submit button is pressed (or whatever interaction symbolizes the execution of the CRUD operation), we should stop there and let each test case decide what state to assert. That gives us more flexibility to use the helpers for testing edge cases or failed states.

dtaylor113 commented on frontend/packages/operator-lifecycle-manager/integration-tests-cypress/tests/operator-install-global.spec.ts:
```diff
@@ -1,131 +1,57 @@
+  deleteUrl: '/api/kubernetes/apis/infinispan.org',
+};
+
+describe(`Globally installing "${testOperator.name}" operator in ${testOperator.installedNamespace}`, () => {
```
Global installs do get installed into a namespace, typically `openshift-operators`, right?

TheRealJon replied:
Yes, but that is generally abstracted from users.

dtaylor113 commented on frontend/packages/operator-lifecycle-manager/integration-tests-cypress/tests/operator-install-global.spec.ts:
```diff
@@ -1,131 +1,57 @@
+  installedNamespace: 'openshift-operators',
+};
+
+const testInstance: TestInstanceProps = {
```
I've seen it refered to both `instances` and `operands` in our code base.

TheRealJon replied:
The correct term is Operand

dtaylor113 commented:
/retest

dtaylor113 commented:
Hi @TheRealJon, I believe I have addressed all your review comments.
Bigger issue might be that we cannot use 'Data Grid' operator, seems to be taking a long time to install during CI -might have to switch to a different Red Hat Operator & Operand.

TheRealJon commented on frontend/packages/operator-lifecycle-manager/integration-tests-cypress/views/operator.ts:
```diff
@@ -0,0 +1,164 @@
+
+    // Install
+    cy.byTestID('install-operator').click();
+    cy.log('verify Operator began installation');
+    cy.byTestID('view-installed-operators-btn').should(
+      'contain',
+      `View installed Operators in Namespace`,
+    );
+    cy.log(`navigate to OperatorHub in Namespace ${installToNamespace}`);
+    cy.byTestID('view-installed-operators-btn').click();
+    cy.byLegacyTestID('resource-title').should('contain', 'Installed Operators');
```
This is still coupled with a successful install. The button you are looking for in these last steps will only appear after a successful installation. I would either move these into the test case or make a separate helper that verifies the successful in-progress steps of the installation.
```suggestion
  },
  installCompleted: () =>
    cy.log('verify Operator began installation');
    cy.byTestID('view-installed-operators-btn').should(
      'contain',
      `View installed Operators in Namespace`,
    );
    cy.log(`navigate to OperatorHub in Namespace ${installToNamespace}`);
    cy.byTestID('view-installed-operators-btn').click();
    cy.byLegacyTestID('resource-title').should('contain', 'Installed Operators');
```

TheRealJon replied:
My mistake, that link shows immediately.

dtaylor113 commented:
/retest
network connection issues

TheRealJon reviewed (approved):
/lgtm

dtaylor113 commented:
/retest

dtaylor113 commented:
/retest

dtaylor113 commented:
/retest

dtaylor113 commented:
Added WIP label while I limit e2e tests to just cypress OLM testing.

dtaylor113 commented:
/retest

dtaylor113 commented:
/retest
1st test passed!

dtaylor113 commented:
/test ci/prow/e2e-gcp-console
1st test passed!

dtaylor113 commented:
/test e2e-gcp-console
1st test passed

dtaylor113 commented:
/test e2e-gcp-console
2nd test passed

dtaylor113 commented:
/retest

dtaylor113 commented:
@spadgett, @TheRealJon, @rhamilto CI seems to be passing using the new Red Hat CodeReady Workspaces
and Red Hat Service Binding operators.  PTAL -thanks

dtaylor113 commented:
Added WIP label, noticing some flakes :-(  Navigation doesn't seem to happen consistently.

dtaylor113 commented:
/test ci/prow/e2e-gcp-console

dtaylor113 commented:
/test e2e-gcp-console

dtaylor113 commented:
/test e2e-gcp-console

rhamilto reviewed (approved):
LGTM.  Will let @spadgett provide final approval.

spadgett commented on frontend/packages/operator-lifecycle-manager/integration-tests-cypress/tests/operator-install-single-namespace.spec.ts:
```diff
@@ -1,125 +1,80 @@
+    projectDropdown.selectProject(GlobalInstalledNamespace);
+    projectDropdown.shouldContain(GlobalInstalledNamespace);
+    cy.get('.loading-skeleton--table').should('not.exist');
+    // eslint-disable-next-line promise/catch-or-return
```
We should look at disabling this rule for the integration tests folder.

spadgett reviewed (approved):
/lgtm

spadgett commented:
/cherry-pick release-4.8
